File: v8-core/src/oracle/taxonomy.rs
//! Controlled Oracle taxonomy and refusal vocabulary (TARGET_ORACLE_SPEC §2,
//! §8, §16.2, Appendix C).

#![allow(dead_code)]

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, Eq, PartialEq, Serialize, Deserialize)]
pub enum OracleRole {
    Parity,
    Hindsight,
    Target,
}

#[derive(Debug, Clone, Copy, Eq, PartialEq, Serialize, Deserialize)]
pub enum AuthorityLevel {
    L1,
    L2,
    L3,
    LiveReceipt,
}

#[derive(Debug, Clone, Copy, Eq, PartialEq, Serialize, Deserialize)]
pub enum Identifiability {
    Identified,
    PartiallyIdentified,
    ModelDerived,
    NotIdentifiable,
}

#[derive(Debug, Clone, Copy, Eq, PartialEq, Serialize, Deserialize)]
pub enum ValueNotion {
    Retrospective,
    Replication,
    ProspectiveShadow,
    LiveRealized,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub struct OracleContext {
    pub role: OracleRole,
    pub authority: AuthorityLevel,
    pub information_contract_id: String,
    pub opportunity_universe_id: String,
    pub utility_contract_id: String,
    pub policy_class_id: String,
    pub cost_model_id: String,
    pub capacity_model_id: String,
    pub environment_target_id: String,
}

/// Canonical fail-closed vocabulary.  Callers receive a typed refusal rather
/// than an ad-hoc string so later O2-O3 surfaces cannot reclassify failures.
#[derive(Debug, Clone, Copy, Eq, PartialEq, Serialize, Deserialize)]
pub enum OracleRefusal {
    MissingDecisionTimeData,
    OutOfSupportAction,
    ExecutionAuthorityTooWeak,
    UndefinedFuture,
    NonIdentifiableFill,
    ConstraintInfeasible,
    ProtectedSliceAlreadyConsumed,
    ModelOnlyCounterfactual,
    InsufficientSupport,
}

impl OracleRefusal {
    pub const fn code(self) -> &'static str {
        match self {
            Self::MissingDecisionTimeData => "MISSING_DECISION_TIME_DATA",
            Self::OutOfSupportAction => "OUT_OF_SUPPORT_ACTION",
            Self::ExecutionAuthorityTooWeak => "EXECUTION_AUTHORITY_TOO_WEAK",
            Self::UndefinedFuture => "UNDEFINED_FUTURE",
            Self::NonIdentifiableFill => "NON_IDENTIFIABLE_FILL",
            Self::ConstraintInfeasible => "CONSTRAINT_INFEASIBLE",
            Self::ProtectedSliceAlreadyConsumed => "PROTECTED_SLICE_ALREADY_CONSUMED",
            Self::ModelOnlyCounterfactual => "MODEL_ONLY_COUNTERFACTUAL",
            Self::InsufficientSupport => "INSUFFICIENT_SUPPORT",
        }
    }
}

// ---------------------------------------------------------------------------
// Issue #AUD-010: The Four Orthogonal Epistemic and Authority Dimensions
// (V8_CONSTITUTION Rule 12, TARGET_ORACLE_SPEC §7, §16, EVALUATION_EVIDENCE_SYSTEM §1-4)
// ---------------------------------------------------------------------------

/// Axis 1: Code & Implementation Verification Dimension.
#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash, Serialize, Deserialize)]
pub enum VerificationDimension {
    /// Satisfies unit tests, type invariants, and state-machine transitions.
    ContractVerified,
    /// Satisfies D-116 differential parity against independent reference engine.
    ImplementationParity,
    /// Satisfies PIT temporal non-interference and permutation invariance.
    MetamorphicInvariant,
}

impl VerificationDimension {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::ContractVerified => "CONTRACT_VERIFIED",
            Self::ImplementationParity => "IMPLEMENTATION_PARITY",
            Self::MetamorphicInvariant => "METAMORPHIC_INVARIANT",
        }
    }
}

/// Axis 2: Economic Evidence & Promotion Stage (Constitution Rule 12).
#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash, Serialize, Deserialize)]
pub enum EconomicEvidenceStage {
    /// Default state for all unpromoted / uncertified research: NO ECONOMIC CLAIM.
    NoEconomicClaim,
    /// Hindsight / regret analysis proves value recovery is structurally feasible.
    RecoverableWithinClass,
    /// Policy admissible under certified multiple-testing and risk constraints.
    PromotableWithinContract,
    /// Empirically demonstrated edge in non-interfering live paper shadow execution.
    ShadowSupported,
    /// Realized predictive profitability with certified multiple-testing adjustments.
    LiveSupported,
}

impl EconomicEvidenceStage {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::NoEconomicClaim => "NO_ECONOMIC_CLAIM",
            Self::RecoverableWithinClass => "RECOVERABLE_WITHIN_CLASS",
            Self::PromotableWithinContract => "PROMOTABLE_WITHIN_CONTRACT",
            Self::ShadowSupported => "SHADOW_SUPPORTED",
            Self::LiveSupported => "LIVE_SUPPORTED",
        }
    }
}

/// Axis 3: Counterfactual & Microstructure Authority Level.
#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash, Serialize, Deserialize)]
pub enum CounterfactualAuthority {
    /// Fully identified from observable order book / matching engine state.
    Identified,
    /// Bounded by partial identification intervals (e.g. Manski bounds).
    PartiallyIdentified,
    /// Derived from parametric or historical simulation models (uncertified).
    ModelDerived,
    /// Microstructure or queue dynamics cannot be identified from available data.
    NotIdentifiable,
}

impl CounterfactualAuthority {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Identified => "IDENTIFIED",
            Self::PartiallyIdentified => "PARTIALLY_IDENTIFIED",
            Self::ModelDerived => "MODEL_DERIVED",
            Self::NotIdentifiable => "NOT_IDENTIFIABLE",
        }
    }
}

/// Axis 4: Statistical Hypothesis Testing Verdict (arXiv:2607.20093 §4).
#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash, Serialize, Deserialize)]
pub enum StatisticalVerdict {
    /// Statistically significant rejection of the null with FDR/FWER control.
    Supported,
    /// Statistically significant evidence against the edge hypothesis (negative result).
    Refuted,
    /// Insufficient sample size or power to distinguish signal from noise.
    InconclusiveUnderpowered,
}

impl StatisticalVerdict {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Supported => "SUPPORTED",
            Self::Refuted => "REFUTED",
            Self::InconclusiveUnderpowered => "INCONCLUSIVE_UNDERPOWERED",
        }
    }
}

/// Canonical Unknown Reason Code Attribution Surface.
#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash, Serialize, Deserialize)]
pub enum UnknownReasonCode {
    MissingDecisionTimeData,
    NonIdentifiableFill,
    InsufficientSupport,
    ModelOnlyCounterfactual,
    ExecutionAuthorityTooWeak,
    OutOfSupportAction,
    UndefinedFuture,
    ConstraintInfeasible,
    ProtectedSliceAlreadyConsumed,
}

impl UnknownReasonCode {
    pub const fn code(self) -> &'static str {
        match self {
            Self::MissingDecisionTimeData => "MISSING_DECISION_TIME_DATA",
            Self::NonIdentifiableFill => "NON_IDENTIFIABLE_FILL",
            Self::InsufficientSupport => "INSUFFICIENT_SUPPORT",
            Self::ModelOnlyCounterfactual => "MODEL_ONLY_COUNTERFACTUAL",
            Self::ExecutionAuthorityTooWeak => "EXECUTION_AUTHORITY_TOO_WEAK",
            Self::OutOfSupportAction => "OUT_OF_SUPPORT_ACTION",
            Self::UndefinedFuture => "UNDEFINED_FUTURE",
            Self::ConstraintInfeasible => "CONSTRAINT_INFEASIBLE",
            Self::ProtectedSliceAlreadyConsumed => "PROTECTED_SLICE_ALREADY_CONSUMED",
        }
    }
}

/// Authority Error for invalid states or Rule 12 violations.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum AuthorityError {
    AmbiguousTaxonomyState(String),
    UncertifiedEconomicClaim(String),
    InvalidReasonCode(String),
}

impl std::fmt::Display for AuthorityError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::AmbiguousTaxonomyState(s) => write!(f, "Ambiguous taxonomy state: {s}"),
            Self::UncertifiedEconomicClaim(s) => write!(f, "Rule 12 violation (uncertified economic claim): {s}"),
            Self::InvalidReasonCode(s) => write!(f, "Invalid unknown reason code: {s}"),
        }
    }
}

impl std::error::Error for AuthorityError {}

/// Full Orthogonal Product Space State ($I1$ invariant).
#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash, Serialize, Deserialize)]
pub struct AuditState {
    pub verification: VerificationDimension,
    pub economic_stage: EconomicEvidenceStage,
    pub counterfactual_authority: CounterfactualAuthority,
    pub statistical_verdict: StatisticalVerdict,
}

impl AuditState {
    pub fn new(
        verification: VerificationDimension,
        economic_stage: EconomicEvidenceStage,
        counterfactual_authority: CounterfactualAuthority,
        statistical_verdict: StatisticalVerdict,
    ) -> Self {
        Self {
            verification,
            economic_stage,
            counterfactual_authority,
            statistical_verdict,
        }
    }

    /// Enforce Constitution Rule 12: No uncertified economic edge claim.
    pub fn validate_rule12(&self) -> Result<(), AuthorityError> {
        if (self.economic_stage == EconomicEvidenceStage::LiveSupported
            || self.economic_stage == EconomicEvidenceStage::ShadowSupported)
            && (self.counterfactual_authority == CounterfactualAuthority::ModelDerived
                || self.counterfactual_authority == CounterfactualAuthority::NotIdentifiable
                || self.statistical_verdict != StatisticalVerdict::Supported)
        {
            return Err(AuthorityError::UncertifiedEconomicClaim(format!(
                "Cannot claim {:?} when authority is {:?} and statistical verdict is {:?}",
                self.economic_stage, self.counterfactual_authority, self.statistical_verdict
            )));
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn taxonomy_has_only_the_three_oracle_roles() {
        let roles = [
            OracleRole::Parity,
            OracleRole::Hindsight,
            OracleRole::Target,
        ];
        assert_eq!(roles.len(), 3);
        assert_ne!(OracleRole::Parity, OracleRole::Target);
        assert_eq!(AuthorityLevel::L1 as u8, 0);
        assert_eq!(ValueNotion::Retrospective as u8, 0);
    }

    #[test]
    fn refusal_codes_are_closed_and_stable() {
        assert_eq!(
            OracleRefusal::MissingDecisionTimeData.code(),
            "MISSING_DECISION_TIME_DATA"
        );
        assert_eq!(
            OracleRefusal::ConstraintInfeasible.code(),
            "CONSTRAINT_INFEASIBLE"
        );
    }
}

File: v8-next/src/v8_next/oracle/taxonomy.py
"""Controlled oracle taxonomy and refusal vocabulary — port of v8-core/src/oracle/taxonomy.rs.

Public vocabulary is closed: the refusal codes and the four orthogonal dimensions are the Rust
ones, spelled the same, so a report from either system reads identically. Note that the Rust
tree carries two distinct types named ``CounterfactualAuthority`` (an enum in taxonomy.rs, a
struct in authority.rs); this port keeps the same split, and the package exports the enum as
``CounterfactualAuthorityLevel`` to avoid the ambiguity at the boundary.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum, StrEnum


class OracleRole(StrEnum):
    PARITY = "PARITY"
    HINDSIGHT = "HINDSIGHT"
    TARGET = "TARGET"


class AuthorityLevel(IntEnum):
    """Ordered: an output may not exceed the authority of its inputs."""

    L1 = 0
    L2 = 1
    L3 = 2
    LIVE_RECEIPT = 3


class Identifiability(StrEnum):
    IDENTIFIED = "IDENTIFIED"
    PARTIALLY_IDENTIFIED = "PARTIALLY_IDENTIFIED"
    MODEL_DERIVED = "MODEL_DERIVED"
    NOT_IDENTIFIABLE = "NOT_IDENTIFIABLE"


class ValueNotion(IntEnum):
    RETROSPECTIVE = 0
    REPLICATION = 1
    PROSPECTIVE_SHADOW = 2
    LIVE_REALIZED = 3


class CounterfactualAuthority(StrEnum):
    """Axis 3: counterfactual and microstructure authority level."""

    IDENTIFIED = "IDENTIFIED"
    PARTIALLY_IDENTIFIED = "PARTIALLY_IDENTIFIED"
    MODEL_DERIVED = "MODEL_DERIVED"
    NOT_IDENTIFIABLE = "NOT_IDENTIFIABLE"


class VerificationDimension(StrEnum):
    """Axis 1: code and implementation verification."""

    CONTRACT_VERIFIED = "CONTRACT_VERIFIED"
    IMPLEMENTATION_PARITY = "IMPLEMENTATION_PARITY"
    METAMORPHIC_INVARIANT = "METAMORPHIC_INVARIANT"


class EconomicEvidenceStage(StrEnum):
    """Axis 2: economic evidence and promotion stage (Constitution Rule 12)."""

    NO_ECONOMIC_CLAIM = "NO_ECONOMIC_CLAIM"
    RECOVERABLE_WITHIN_CLASS = "RECOVERABLE_WITHIN_CLASS"
    PROMOTABLE_WITHIN_CONTRACT = "PROMOTABLE_WITHIN_CONTRACT"
    SHADOW_SUPPORTED = "SHADOW_SUPPORTED"
    LIVE_SUPPORTED = "LIVE_SUPPORTED"


class StatisticalVerdict(StrEnum):
    """Axis 4: statistical hypothesis testing verdict."""

    SUPPORTED = "SUPPORTED"
    REFUTED = "REFUTED"
    INCONCLUSIVE_UNDERPOWERED = "INCONCLUSIVE_UNDERPOWERED"


class OracleRefusal(StrEnum):
    """Canonical fail-closed vocabulary; a refusal is typed, never an ad-hoc string."""

    MISSING_DECISION_TIME_DATA = "MISSING_DECISION_TIME_DATA"
    OUT_OF_SUPPORT_ACTION = "OUT_OF_SUPPORT_ACTION"
    EXECUTION_AUTHORITY_TOO_WEAK = "EXECUTION_AUTHORITY_TOO_WEAK"
    UNDEFINED_FUTURE = "UNDEFINED_FUTURE"
    NON_IDENTIFIABLE_FILL = "NON_IDENTIFIABLE_FILL"
    CONSTRAINT_INFEASIBLE = "CONSTRAINT_INFEASIBLE"
    PROTECTED_SLICE_ALREADY_CONSUMED = "PROTECTED_SLICE_ALREADY_CONSUMED"
    MODEL_ONLY_COUNTERFACTUAL = "MODEL_ONLY_COUNTERFACTUAL"
    INSUFFICIENT_SUPPORT = "INSUFFICIENT_SUPPORT"

    def code(self) -> str:
        return self.value


#: the same vocabulary, seen from the "why is this unknown" side
UnknownReasonCode = OracleRefusal


@dataclass(frozen=True)
class OracleContext:
    """The declared context a counterfactual is evaluated inside."""

    role: OracleRole
    authority: AuthorityLevel
    information_contract_id: str
    opportunity_universe_id: str
    utility_contract_id: str
    policy_class_id: str
    cost_model_id: str
    capacity_model_id: str
    environment_target_id: str


class AuthorityError(Exception):
    """Invalid taxonomy state or a Rule 12 violation."""

    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code


@dataclass(frozen=True)
class AuditState:
    """The full orthogonal product-space state (I1 invariant)."""

    verification: VerificationDimension
    economic_stage: EconomicEvidenceStage
    counterfactual_authority: CounterfactualAuthority
    statistical_verdict: StatisticalVerdict

    def validate_rule12(self) -> None:
        """Constitution Rule 12: no uncertified economic edge claim.

        A supported or live-supported stage may not rest on a model-derived or
        not-identifiable counterfactual, nor on a verdict that is not supported.
        """
        claims_edge = self.economic_stage in (
            EconomicEvidenceStage.SHADOW_SUPPORTED,
            EconomicEvidenceStage.LIVE_SUPPORTED,
        )
        weak_authority = self.counterfactual_authority in (
            CounterfactualAuthority.MODEL_DERIVED,
            CounterfactualAuthority.NOT_IDENTIFIABLE,
        )
        if claims_edge and (weak_authority or self.statistical_verdict is not StatisticalVerdict.SUPPORTED):
            raise AuthorityError(
                "UNCERTIFIED_ECONOMIC_CLAIM",
                f"Cannot claim {self.economic_stage.value} when authority is "
                f"{self.counterfactual_authority.value} and statistical verdict is "
                f"{self.statistical_verdict.value}",
            )
